### PR TITLE
alternator/executor: Use native reversed format

### PR DESCRIPTION
**Pull request by paszkow**
When executing reversed queries, a native revered format shall be used. Therefore, the table schema and the clustering key bounds are reversed before a partition slice and a read command are constructed.

It is, however, possible to run a reversed query passing a table schema but only when there are no restrictions on the clustering keys. In this particular situation, the query returns correct results. Since the current alternator tests in test.py do not imply any restrictions, this situation was not caught during development of https://github.com/scylladb/scylladb/pull/18864.

Hence, additional tests are provided that add clustering keys restrictions when executing reversed queries to capture such errors earlier than in dtests.

Additional manual tests were performed to test a mixed-node cluster (with alternator API enabled in Scylla on each node):

1. 2-node cluster with one node upgraded: reverse read queries performed on an old node
2. 2-node cluster with one node upgraded: reverse read queries performed on a new node
3. 2-node cluster with one node upgraded and all its sstable files deleted to trigger repair: reverse read queries performed on an old node
4. 2-node cluster with one node upgraded and all its sstable files deleted to trigger repair: reverse read queries performed on a new node

All reverse read queries above consists of:

- single-partition reverse reads with no clustering key restrictions, with single column restrictions and multi column restrictions both with and without paging turned on

The exact same tests were also performed on a fully upgraded cluster.

Fixes https://github.com/scylladb/scylladb/issues/20191

No backport is required as this is a complementary patch for the series https://github.com/scylladb/scylladb/pull/18864 that did not require backporting.

### CODE REVIEW

File: alternator/executor.cc [hunk 1]
```diff
@@ -3816,6 +3816,17 @@ static future<executor::request_return_type> do_query(service::storage_proxy& pr
     auto selection = cql3::selection::selection::wildcard(schema);
     query::partition_slice::option_set opts = selection->get_query_options();
     opts.add(custom_opts);
+
+    // Reverse the schema and the clustering bounds as the underlying code expects
+    // reversed queries in the native reversed format.
+    const bool reversed = opts.contains<query::partition_slice::option::reversed>();
```
Comment (bhalevy) [inline alternator/executor.cc:3822]: isn't that the legacy option?
how does that work during upgrade?

Comment (paszkow) [reply alternator/executor.cc:3822]: No. This option is to distinguish between forward and reversed partition slices. It is set to true no matter of the format, i.e. legacy or native. Now, we always require the reversed queries to be in the native reverse format. 

Storage proxy is responsible for translating queries from native format into legacy one in case of mixed-node clusters. There is a feature flag `NATIVE_REVERSE_QUERIES` used for this purpose.

**Review by bhalevy — APPROVED** [review]

**Comment by bhalevy** [comment]
LGTM.

@nyh / @denesb please review and consider merging ASAP.

**Comment by bhalevy** [comment]
@paszkow I presume that you've tested this fix with the failing dtests.
Can you link a BYO job that tests that so we have it on record?

**Comment by nyh** [comment]
@bhalevy I'm on vacation

**Comment by nyh** [comment]
There's one thing I can say though: If this is a regression caught only by a complex slow test, please and can be reproduced by a smaller quick single-node test, please consider adding a small regression test in test/alternator. Also please consider how the dtest did not prevent the regression going in - was it not in "gating"? Should it?

**Comment by nyh** [comment]
There's one thing I can say though: If this is a regression caught only by a complex slow test, please and can be reproduced by a smaller quick single-node test, please consider adding a small regression test in test/alternator. Also please consider how the dtest did not prevent the regression going in - was it not in "gating"? Should it?

**Comment by nyh** [comment]
> > @paszkow I presume that you've tested this fix with the failing dtests. Can you link a BYO job that tests that so we have it on record?
> 
> I run it locally only. Do not rush with merging. Now unit tests fail. Looking at them now.

To run the entire Alternator functional test suite, just build Scylla and run test/alternator/run. It takes less than 30 seconds for the entire test suite and it's easy to do locally.

**Comment by paszkow** [comment]
> > > @paszkow I presume that you've tested this fix with the failing dtests. Can you link a BYO job that tests that so we have it on record?
> > 
> > 
> > I run it locally only. Do not rush with merging. Now unit tests fail. Looking at them now.
> 
> To run the entire Alternator functional test suite, just build Scylla and run test/alternator/run. It takes less than 30 seconds for the entire test suite and it's easy to do locally.

Sure. Running the tests is not an issue. I am experiencing some very strange behaviour. Whenever I fix dtests, unit tests fail and the other way around. I am missing something, I cannot see at the moment. Hence, changed the PR into draft, so it does not get accidentally merged.

File: test/alternator/test_query.py [hunk 2]
```diff
@@ -427,7 +427,7 @@ def test_query_reverse_paging(test_table_sn):
     # at a time and resuming the query.
     for limit in [1, 2, 3, 7, 10, 17, 100, 10000]:
         got_items = full_query(test_table_sn, KeyConditions={'p': {'AttributeValueList': [p], 'ComparisonOperator': 'EQ'}}, ScanIndexForward=False, Limit=limit)
-        got_sort_keys = [x['c'] for x in got_items]
+        got_sort_keys = [Decimal(x['c']) for x in got_items]
```
Comment (denesb) [inline test/alternator/test_query.py:430]: Why this change?

**Review by denesb — COMMENTED** [review]
I cannot determine whether this PR is correct, because I know almost nothing about how Alternator's front-end works.
This will need a review from @nyh, when he is back from holiday.

File: test/alternator/test_query.py (hunk 2 above)
Comment (paszkow) [reply test/alternator/test_query.py]: Leftover from my approaches.

**Comment by paszkow** [comment]
v2:

- Correctly reverse schema and use a query schema only when required
- Rename schema into table_schema in `alternator::do_query` for better readability

@bhalevy Run all alternator related unit tests and dtests for the release and the debug modes. All passed. Launched BYO to run these tests again https://jenkins.scylladb.com/job/scylla-master/job/byo/job/byo_build_tests_dtest/2452/

**Comment by paszkow** [comment]
> #### Failed Tests (3/32257):
> * [test_safe_unsafe_rebuild[1]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11019/testReport/junit/repair_based_node_operations_test/TestRepairBasedNodeOperations/Tests___dtest___test_safe_unsafe_rebuild_1_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_safe_unsafe_rebuild%5B1%5D)
> * [test_safe_unsafe_rebuild[1]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11019/testReport/junit/repair_based_node_operations_test/TestRepairBasedNodeOperations/Tests___dtest_with_topology_changes___test_safe_unsafe_rebuild_1_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_safe_unsafe_rebuild%5B1%5D)
> * [test_safe_unsafe_rebuild[1]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11019/testReport/junit/repair_based_node_operations_test/TestRepairBasedNodeOperations/Tests___dtest_with_topology_changes___test_safe_unsafe_rebuild_1__2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_safe_unsafe_rebuild%5B1%5D)

The failed tests are unrelated to the changes. They fail on the current master as well. The BYO executed to run all alternator's related unit tests and dtests succeeded, https://jenkins.scylladb.com/job/scylla-master/job/byo/job/byo_build_tests_dtest/2453/

**Comment by paszkow** [comment]
v3:

- Added tests to test.py to run reversed queries with clustering key restrictions

**Comment by avikivity** [comment]
Series looks good. But I need a statement in the cover letter that explains what manual tests took place to prove mixed-version clusters are safe.

**Comment by bhalevy** [comment]
> Series looks good. But I need a statement in the cover letter that explains what manual tests took place to prove mixed-version clusters are safe.

ping @paszkow please add the required information to the cover letter.

**Comment by paszkow** [comment]
> Series looks good. But I need a statement in the cover letter that explains what manual tests took place to prove mixed-version clusters are safe.

Tested and the missing information has been added.
